Sage/carbon
# feat(dialog): expose focus method of root dom node via a ref handle (FE-6176)

addresses #6250

### Proposed behaviour

- Allow ability for consumers to forward a ref handle to `Dialog`, which exposes the component's root DOM node `focus()` method:

```tsx
import Dialog from "carbon-react/lib/components/dialog";
import type { DialogHandle } from "carbon-react/lib/components/dialog";

const MyComponent = () => {
     const dialogHandle = useRef<DialogHandle>(null);

     return (
         <Dialog ref={dialogHandle}>
             <Button onClick={() => dialogHandle.current?.focus()}>
                  Refocus on dialog
             </Button>
         </Dialog>
     );
};
``` 

- Update `Dialog` tests to address DOM elements, particularly those used by React Portals, that were not cleaned-up after tests

### Current behaviour

Currently, there is no way to programmatically refocus on the `Dialog`'s root DOM node. This functionality is required for cases where the content inside the `Dialog` is dynamically updated and screen reader users need to be informed about this change.

### Checklist




#### QA


### Additional context

The purposed changes only exposes the `focus()` method of `Dialog`'s root DOM node, as opposed to the whole DOM node. This has been done to intentionally prevent unnecessary access to the other methods of the root DOM node, as providing access to these methods could make support more difficult. 

### Testing instructions

Using Safari with macOS VoiceOver, Chrome with NVDA and Firefox with NVDA. Please do the following:

1. Do `npm start`. Once storybook is loaded, go to the following URL in the browser: 
```
http://localhost:9001/iframe.html?args=&id=dialog--using-handle&viewMode=story
```
2. Start screen reader
3. Open the dialog, then navigate with the keyboard to the submit button and submit form. This should update the dialog's content.
4. Verify screen reader announces the new content after form submission.
